lib/linter/apply-disable-directives.js

```diff
@@ -197,62 +197,93 @@ function processUnusedDisableDirectives(allDirectives) {
  * for the exported function, except that `reportUnusedDisableDirectives` is not supported
  * (this function always reports unused disable directives).
  * @returns {{problems: Problem[], unusedDisableDirectives: Problem[]}} An object with a list
- * of filtered problems and unused eslint-disable directives
+ * of problems (including suppressed ones) and unused eslint-disable directives
  */
 function applyDirectives(options) {
     const problems = [];
-    let nextDirectiveIndex = 0;
     let currentGlobalDisableDirective = null;
-    const disabledRuleMap = new Map();
 
-    // enabledRules is only used when there is a current global disable directive.
-    const enabledRules = new Set();
     const usedDisableDirectives = new Set();
 
     for (const problem of options.problems) {
+        let nextDirectiveIndex = 0;
+
+        // isProblemSuppressedBefore indicates whether the problem is suppressed by block directives.
+        const isProblemSuppressedBefore = problem.suppressions ? true : false; // eslint-disable-line no-unneeded-ternary -- problem.suppressions might be changed afterwards.
+
+        const usedDisableDirectivesForProblem = new Set();
+        let suppressions = [];
+
         while (
             nextDirectiveIndex < options.directives.length &&
             compareLocations(options.directives[nextDirectiveIndex], problem) <= 0
         ) {
             const directive = options.directives[nextDirectiveIndex++];
+            const suppression = { kind: "directive", justification: directive.unprocessedDirective.justification };
 
             switch (directive.type) {
                 case "disable":
-                    if (directive.ruleId === null) {
-                        currentGlobalDisableDirective = directive;
-                        disabledRuleMap.clear();
-                        enabledRules.clear();
-                    } else if (currentGlobalDisableDirective) {
-                        enabledRules.delete(directive.ruleId);
-                        disabledRuleMap.set(directive.ruleId, directive);
-                    } else {
-                        disabledRuleMap.set(directive.ruleId, directive);
+                    if (!isProblemSuppressedBefore) {
+                        if (directive.ruleId === null) {
+                            currentGlobalDisableDirective = directive;
+                        } else if (directive.ruleId === problem.ruleId && !currentGlobalDisableDirective && suppressions.length === 0) {
```

Comment:
This changes behavior in regard to which directive will be considered used when there are multiple directives disabling the same problem.

For example, if this is `foo.js`:

```js
/* eslint no-undef: 2 */

/* eslint-disable no-undef */

/* eslint-disable no-undef */

foo;

```

on the `main` branch, directive that appears later is considered used, and the other one is reported as unused:

```
D:\projects\eslint>npx eslint foo.js --report-unused-disable-directives

D:\projects\eslint\foo.js
  3:1  error  Unused eslint-disable directive (no problems were reported from 'no-undef')

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.


D:\projects\eslint>
```

on this branch, it's the opposite:

```
D:\tmp\suppressions\eslint>npx eslint foo.js --report-unused-disable-directives

D:\tmp\suppressions\eslint\foo.js
  5:1  error  Unused eslint-disable directive (no problems were reported from 'no-undef')

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.


D:\tmp\suppressions\eslint>
```

I think we should keep the behavior from the `main` branch, as it's usually preferable to remove directive with a broader range.

Reply:
This can be fixed by https://github.com/eslint/eslint/pull/15459#discussion_r785479181